File: playwright-csharp/tests/Data/DataBuilders/RestfulBooker/BookingDataBuilder.cs
using System.Globalization;
using System.Text.Json.Nodes;
using TestTracks.Playwright.CSharp.Specs.Data;

namespace TestTracks.Playwright.CSharp.Specs.Data.DataBuilders.RestfulBooker;

public sealed record BookingDataDefaults
{
    public required JsonObject Payload { get; init; }

    public int CheckInDaysFromToday { get; init; }

    public int CheckOutDaysFromToday { get; init; }

    public int FirstNameMaxLength { get; init; } = 18;

    public int LastNameMaxLength { get; init; } = 30;
}

public sealed class BookingDataBuilder
{
    private const int NewBookingId = 0;
    private readonly JsonObject _payload;

    public BookingDataBuilder(BookingDataDefaults defaults)
    {
        _payload = defaults.Payload.DeepClone().AsObject();
        SetBookingDates(defaults.CheckInDaysFromToday, defaults.CheckOutDaysFromToday);
    }

    public static BookingDataBuilder ForScenario(string scenarioId, BookingDataDefaults defaults)
    {
        var builder = new BookingDataBuilder(defaults);
        var firstName = builder.StringValue("firstname");
        var lastName = $"{builder.StringValue("lastname")}{scenarioId}";
        var email = builder.StringValue("email");
        var emailParts = email.Split('@', 2);
        var emailPrefix = emailParts[0];
        var emailDomain = emailParts.Length == 2 ? emailParts[1] : "example.test";

        return builder
            .With("firstname", TestData.SafeName(firstName, scenarioId, defaults.FirstNameMaxLength))
            .With("lastname", lastName[..Math.Min(defaults.LastNameMaxLength, lastName.Length)])
            .With("email", $"{emailPrefix}.{scenarioId}@{emailDomain}");
    }

    public BookingDataBuilder ForRoom(int roomId)
    {
        return With("roomid", roomId);
    }

    public BookingDataBuilder With(string propertyName, object? value)
    {
        _payload[propertyName] = JsonValue.Create(value);
        return this;
    }

    public JsonObject BuildPayload()
    {
        return _payload.DeepClone().AsObject();
    }

    public BookingFormData BuildFormData()
    {
        var dates = _payload["bookingdates"]?.AsObject()
            ?? throw new InvalidOperationException("Booking payload is missing 'bookingdates'.");

        return new BookingFormData(
            NewBookingId,
            IntValue("roomid"),
            StringValue("firstname"),
            StringValue("lastname"),
            BoolValue("depositpaid"),
            StringValue("email"),
            StringValue("phone"),
            new BookingDateRange(
                DateOnly.Parse(StringValue(dates, "checkin"), CultureInfo.InvariantCulture),
                DateOnly.Parse(StringValue(dates, "checkout"), CultureInfo.InvariantCulture)));
    }

    private void SetBookingDates(int checkInDaysFromToday, int checkOutDaysFromToday)
    {
        var bookingDates = _payload["bookingdates"] as JsonObject;
        if (bookingDates is null)
        {
            bookingDates = new JsonObject();
            _payload["bookingdates"] = bookingDates;
        }

        bookingDates["checkin"] = DateOnly.FromDateTime(DateTime.UtcNow.Date.AddDays(checkInDaysFromToday))
            .ToString("yyyy-MM-dd", CultureInfo.InvariantCulture);
        bookingDates["checkout"] = DateOnly.FromDateTime(DateTime.UtcNow.Date.AddDays(checkOutDaysFromToday))
            .ToString("yyyy-MM-dd", CultureInfo.InvariantCulture);
    }

    private string StringValue(string propertyName)
    {
        return StringValue(_payload, propertyName);
    }

    private static string StringValue(JsonObject payload, string propertyName)
    {
        return payload[propertyName]?.GetValue<string>()
            ?? throw new InvalidOperationException($"Booking payload is missing '{propertyName}'.");
    }

    private int IntValue(string propertyName)
    {
        return _payload[propertyName]?.GetValue<int>()
            ?? throw new InvalidOperationException($"Booking payload is missing '{propertyName}'.");
    }

    private bool BoolValue(string propertyName)
    {
        return _payload[propertyName]?.GetValue<bool>()
            ?? throw new InvalidOperationException($"Booking payload is missing '{propertyName}'.");
    }
}

public sealed record BookingFormData(
    int BookingId,
    int RoomId,
    string FirstName,
    string LastName,
    bool DepositPaid,
    string Email,
    string Phone,
    BookingDateRange BookingDates);

public sealed record BookingDateRange(DateOnly CheckIn, DateOnly CheckOut);

File: playwright-java/src/test/java/testtracks/data/builders/restfulbooker/BookingDataBuilder.java
package testtracks.data.builders.restfulbooker;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.JsonNodeFactory;
import com.fasterxml.jackson.databind.node.ObjectNode;
import java.time.LocalDate;
import testtracks.data.JsonPayloads;
import testtracks.data.TestData;

public final class BookingDataBuilder {
  private static final int NEW_BOOKING_ID = 0;
  private static final ObjectMapper JSON = new ObjectMapper();
  private final ObjectNode payload;

  public BookingDataBuilder(BookingDataDefaults defaults) {
    payload = defaults.payload().deepCopy();
    setBookingDates(defaults.checkInDaysFromToday(), defaults.checkOutDaysFromToday());
  }

  public static BookingDataBuilder forScenario(String scenarioId, BookingDataDefaults defaults) {
    var builder = new BookingDataBuilder(defaults);
    var firstName = builder.stringValue("firstname");
    var lastName = builder.stringValue("lastname") + scenarioId;
    var email = builder.stringValue("email");
    var emailParts = email.split("@", 2);
    var emailPrefix = emailParts[0];
    var emailDomain = emailParts.length == 2 ? emailParts[1] : "example.test";

    return builder
        .with("firstname", TestData.safeName(firstName, scenarioId, defaults.firstNameMaxLength()))
        .with(
            "lastname",
            lastName.substring(0, Math.min(defaults.lastNameMaxLength(), lastName.length())))
        .with("email", emailPrefix + "." + scenarioId + "@" + emailDomain);
  }

  public BookingDataBuilder forRoom(int roomId) {
    return with("roomid", roomId);
  }

  public BookingDataBuilder with(String propertyName, Object value) {
    payload.set(propertyName, valueToJson(value));
    return this;
  }

  public ObjectNode buildPayload() {
    return payload.deepCopy();
  }

  public BookingFormData buildFormData() {
    var dates = payload.get("bookingdates");
    if (!(dates instanceof ObjectNode bookingDates)) {
      throw new IllegalStateException("Booking payload is missing 'bookingdates'.");
    }

    return new BookingFormData(
        NEW_BOOKING_ID,
        intValue("roomid"),
        stringValue("firstname"),
        stringValue("lastname"),
        boolValue("depositpaid"),
        stringValue("email"),
        stringValue("phone"),
        new BookingDateRange(
            LocalDate.parse(JsonPayloads.stringValue(bookingDates, "checkin")),
            LocalDate.parse(JsonPayloads.stringValue(bookingDates, "checkout"))));
  }

  private void setBookingDates(int checkInDaysFromToday, int checkOutDaysFromToday) {
    ObjectNode bookingDates;
    if (payload.get("bookingdates") instanceof ObjectNode existing) {
      bookingDates = existing;
    } else {
      bookingDates = JsonNodeFactory.instance.objectNode();
      payload.set("bookingdates", bookingDates);
    }

    bookingDates.put("checkin", LocalDate.now().plusDays(checkInDaysFromToday).toString());
    bookingDates.put("checkout", LocalDate.now().plusDays(checkOutDaysFromToday).toString());
  }

  private String stringValue(String propertyName) {
    return JsonPayloads.stringValue(payload, propertyName);
  }

  private int intValue(String propertyName) {
    return JsonPayloads.intValue(payload, propertyName);
  }

  private boolean boolValue(String propertyName) {
    var value = payload.get(propertyName);
    if (value == null || value.isNull()) {
      throw new IllegalStateException("Booking payload is missing '" + propertyName + "'.");
    }
    return value.asBoolean();
  }

  private static JsonNode valueToJson(Object value) {
    return JSON.valueToTree(value);
  }
}
